### PR TITLE
Ensure adaptive scaling is properly awaited and closed

### PR DESCRIPTION
**Pull request by fjetter**
* Primarily this ensures that scale_up/down is properly awaited in case the `sync` method is used which is more common than the `loop.add_callback`.
* It also ensures the adaptive PC is stopped properly once a cluster shuts down.
* A bit cleanup. There has been a test which raises error logs and I had issues figuring out what it actually was doing. It also was refactored a few times since its original inception.

### CODE REVIEW

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -713,7 +713,7 @@ async def test_scheduler_sees_memory_limits(s):
     await w.close()
 
 
-@gen_cluster(client=True, timeout=1000)
+@gen_cluster(client=True)
```
Comment (fjetter) [inline distributed/tests/test_scheduler.py:716]: This is an unrelated change which slipped in. This test usually runs a second. If we run into a timeout, I'd like to know since this may hint towards an issue

Comment (fjetter) [reply distributed/tests/test_scheduler.py:716]: If desired I'll remove this. I'll follow up with a few things related to worker closing/retiring later on and can include it there

File: distributed/deploy/tests/test_adaptive.py [hunk 2]
```diff
@@ -19,43 +20,6 @@
 )
 
 
-@pytest.mark.asyncio
-async def test_simultaneous_scale_up_and_down(cleanup):
```
Comment (fjetter) [inline distributed/deploy/tests/test_adaptive.py:23]: This test was added in https://github.com/dask/distributed/pull/1608 to disallow simultaneous up-/down-scaling but the current implementation doesn't work at all since the API changed. Instead it only raises error logs

File: distributed/deploy/cluster.py [hunk 3]
```diff
@@ -82,6 +83,8 @@ async def _close(self):
         if self.status == Status.closed:
             return
 
+        self._adaptive.stop()
```
Comment (jrbourbeau) [inline distributed/deploy/cluster.py:86]: Since `self._adaptive` is only set after `Cluster.adapt` is called, we might want to add a `with suppress(AttributeError):` here

**Comment by fjetter** [comment]
Test failure is #4651 / #4719

File: distributed/deploy/tests/test_adaptive.py [hunk 4]
```diff
@@ -460,3 +424,58 @@ async def test_update_adaptive(cleanup):
         await asyncio.sleep(0.2)
         assert first.periodic_callback is None
         assert second.periodic_callback.is_running()
+
+
+from distributed.utils_test import slowinc
```
Comment (jrbourbeau) [inline distributed/deploy/tests/test_adaptive.py:429]: This appears to be already imported at the top of this module

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (jrbourbeau) [reply distributed/tests/test_scheduler.py:719]: +1 for removing

File: distributed/deploy/tests/test_adaptive.py [hunk 5]
```diff
@@ -460,3 +424,58 @@ async def test_update_adaptive(cleanup):
...
+            cluster.adapt()
+
+            await client.gather(futures)
+
+            del futures
+            await async_wait_for(lambda: not cluster.workers, 10)
+
+
+@pytest.mark.asyncio
+async def test_adaptive_stopped():
```
Comment (jrbourbeau) [inline distributed/deploy/tests/test_adaptive.py:462]: Given that this uses an async cluster, do we need the `async_wait_for` + timeout or can we just `assert` the various attributes directly? For example, when I make the following changes locally this test still passes:

```diff
diff --git a/distributed/deploy/tests/test_adaptive.py b/distributed/deploy/tests/test_adaptive.py
index 39284302..8cf4f381 100644
--- a/distributed/deploy/tests/test_adaptive.py
+++ b/distributed/deploy/tests/test_adaptive.py
@@ -466,16 +466,8 @@ async def test_adaptive_stopped():
     """
     async with LocalCluster(n_workers=0, asynchronous=True) as cluster:
         async with Client(cluster, asynchronous=True) as client:
-            instance = cluster.adapt(interval="10ms")
+            pc = cluster.adapt(interval="10ms").periodic_callback
+            assert pc is not None
+            assert pc.is_running() is not None

-            await async_wait_for(
-                lambda: instance.periodic_callback is not None, timeout=5
-            )
-
-            await async_wait_for(
-                lambda: instance.periodic_callback.is_running() is not None, timeout=5
-            )
-
-            pc = instance.periodic_callback
-
-    await async_wait_for(lambda: pc.is_running() is not None, timeout=5)
+    assert pc.is_running() is not None
```

**Review by jrbourbeau — COMMENTED** [review]
Thanks @fjetter! I left a few small comments, but overall the changes here look good to me

File: distributed/deploy/tests/test_adaptive.py (hunk 5 above)
Comment (fjetter) [reply distributed/deploy/tests/test_adaptive.py:462]: Well, at the very least we'll need to have one wait in between which actually waits for the entire thing to start. If the PC was never started, the conditions are trivially true. I'll add a comment and what I can remove for it to still work

File: distributed/deploy/tests/test_adaptive.py (hunk 4 above)
Comment (fjetter) [reply distributed/deploy/tests/test_adaptive.py:429]: Wondering why me pre-commit didn't catch that... will fix of course

File: distributed/deploy/tests/test_adaptive.py (hunk 5 above)
Comment (fjetter) [reply distributed/deploy/tests/test_adaptive.py:459]: All of the `is not None` checks where useless which is why the test did not fail. I properly assert for bool now and now I need to wait

File: distributed/deploy/spec.py [hunk 6]
```diff
@@ -413,7 +414,15 @@ async def _close(self):
             return
         if self.status == Status.running or self.status == Status.failed:
             self.status = Status.closing
-            self.scale(0)
+
+            # Need to call stop here before we close all servers to avoid having
+            # dangling tasks in the ioloop
+            with suppress(AttributeError):
+                self._adaptive.stop()
```
Comment (fjetter) [inline distributed/deploy/spec.py:421]: This is a bit unfortunate since I need to call it up here and not down in `Cluster`. If I do this only in cluster, the even loop seems to close too soon and we have still pending tasks from `AdaptiveCore.adapt`.

I'm wondering if we ever considered adding `PYTHONASYNCIODEBUG=1` to our test suite which would raise in these instances. Not sure how much would break or if this is a bad idea in general
